### PR TITLE
Validate Project-URL metadata

### PR DESCRIPTION
**Pull request by ewdurbin**
RE: #2666

### CODE REVIEW

File: warehouse/forklift/legacy.py [hunk 1]
```diff
@@ -251,8 +251,11 @@ def _validate_project_url(value):
 
 
 def _validate_project_url_list(form, field):
-    for datum in field.data:
-        _validate_project_url(datum)
+    if isinstance(field.data, str):
+        _validate_project_url(field.data)  # Or raise ValidationError?
```
Comment (ewdurbin) [inline warehouse/forklift/legacy.py]: @dstufft i'm not sure if we want to just accept a string or not here.

Comment (di) [reply warehouse/forklift/legacy.py]: `Project-URL` is a multiple-use field, so this should always be a list. I think it should just fail validation if it isn't.

Comment (ewdurbin) [reply warehouse/forklift/legacy.py]: oh, you're darn tootin.

https://github.com/pypa/warehouse/blob/6fe519b7506702bdf961631bdb2373670842bf21/warehouse/forklift/legacy.py#L464-L469

**Comment by di** [comment]
Wouldn't this be an issue for any `ListField`-type field, not just `project_urls`?

**Comment by ewdurbin** [comment]
yeah, let's see if I can put a validator on the `ListField` itself, just now realized we own that class :-D

**Comment by di** [comment]
I think we might want a failing test like this on `TestListField`:

```python
@pytest.mark.parametrize(
    ("value", "expected"),
    [
        ("", []),
        ("foobar", ["foobar"]),
    ]
)
def test_coerce_string_into_list(self, value, expected):
    class MyForm(Form):
        test = legacy.ListField()

    form = MyForm(MultiDict({'test': value}))

    assert form.test.data == expected
```

The issue seems to be that `process_formdata` isn't throwing away values that end up as empty strings.

**Comment by ewdurbin** [comment]
yep! I'm retooling to give `ListField` a `pre_validate` method that should do the job nicely.

File: warehouse/forklift/legacy.py [hunk 2]
```diff
@@ -264,7 +264,7 @@ def _construct_dependencies(form, types):
 class ListField(wtforms.Field):
 
     def process_formdata(self, valuelist):
-        self.data = [v.strip() for v in valuelist]
+        self.data = [v.strip() for v in valuelist if v]
```
Comment (di) [inline warehouse/forklift/legacy.py]: This still could give us empty strings if one of the values is just whitespace:
```
>>> [v.strip() for v in ['', ' ', 'a'] if v]
['', 'a']
```

**Pull request merged**